nmostesting/GenericTest.py

```diff
@@ -413,6 +457,51 @@ def do_test_404_path(self, api_name):
         else:
             return test.FAIL(message)
 
+    def do_test_authorization(self, api_name, test_name, error_code=401, error_type="invalid_token", token=None):
+        api = self.apis[api_name]
+        url = "{}".format(api["url"].rstrip("/"))
+        test = Test("GET /x-nmos/{}/{} ({})".format(api_name, api["version"], test_name), self.auto_test_name(api_name))
+
+        if self.authorization:
+            headers = {}
+            if token:
+                headers = {"Authorization": "Bearer {}".format(token)}
+            valid, response = self.do_request("GET", url, headers=headers)
+            if not valid:
+                return test.FAIL(response)
+
+            if response.status_code != error_code:
+                return test.FAIL("Incorrect response code, expected {}. Received {}"
+                                 .format(error_code, response.status_code))
+
+            if "WWW-Authenticate" not in response.headers:
+                return test.FAIL("Authorization error responses must include a 'WWW-Authenticate' header")
+            if not response.headers["WWW-Authenticate"].startswith("Bearer "):
+                return test.FAIL("'WWW-Authenticate' response header must begin 'Bearer'")
+
+            # https://tools.ietf.org/html/rfc6750#section-3
+            if error_type is not None:
+                valid, message = self.check_auth_error_response("GET", response, error_type)
+                if not valid:
+                    return test.FAIL(message)
```

Comment:
Could the schema be verified after the header? the header is the RFC verification, which is more standard, followed by the NMOS verification.

Reply:
I may wait until we've resolved https://github.com/AMWA-TV/nmos-authorization/pull/65. I think the header requirements are a 'should', and we try to return 'warnings' after 'fails'. If we choose not to mandate any response body then certainly this will change.

Reply:
>If the protected resource request does not include authentication
   credentials or does not contain an access token that enables access
   to the protected resource, the resource server MUST include the HTTP
   "WWW-Authenticate" response header field;

spec says they must have the header https://tools.ietf.org/html/rfc6750#section-3 but...

> If the protected resource request included an access token and failed
   authentication, the resource server SHOULD include the "error"
   attribute to provide the client with the reason why the access
   request was declined.

Reply:
I think that matches the current way the test works. The header is checked for and a fail issued if not. The next step is (currently) to check the response body which may result in a fail, and then finally the 'error' keyword is checked for which can result in a warning.